flink-yarn/src/main/java/org/apache/flink/yarn/entrypoint/YarnJobClusterEntrypoint.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.metrics.MetricRegistry;
+import org.apache.flink.runtime.resourcemanager.ResourceManager;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerConfiguration;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerRuntimeServices;
+import org.apache.flink.runtime.resourcemanager.ResourceManagerRuntimeServicesConfiguration;
+import org.apache.flink.runtime.rpc.FatalErrorHandler;
+import org.apache.flink.runtime.rpc.RpcService;
+import org.apache.flink.runtime.security.SecurityContext;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.Preconditions;
+import org.apache.flink.yarn.YarnResourceManager;
+
+import org.apache.hadoop.yarn.api.ApplicationConstants;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.util.Map;
+
+public class YarnJobClusterEntrypoint extends JobClusterEntrypoint {
+
+	/** The job graph file path. */
+	public static final String JOB_GRAPH_FILE_PATH = "flink.jobgraph.path";
+
+	private final String workingDirectory;
+
+	public YarnJobClusterEntrypoint(
+			Configuration configuration,
+			String workingDirectory) {
+
+		super(configuration);
+		this.workingDirectory = Preconditions.checkNotNull(workingDirectory);
+	}
+
+	@Override
+	protected SecurityContext installSecurityContext(Configuration configuration) throws Exception {
+		return YarnEntrypointUtils.installSecurityContext(configuration, workingDirectory);
+	}
+
+	@Override
+	protected ResourceManager<?> createResourceManager(
+			Configuration configuration,
+			ResourceID resourceId,
+			RpcService rpcService,
+			HighAvailabilityServices highAvailabilityServices,
+			HeartbeatServices heartbeatServices,
+			MetricRegistry metricRegistry,
+			FatalErrorHandler fatalErrorHandler) throws Exception {
+		final ResourceManagerConfiguration rmConfiguration = ResourceManagerConfiguration.fromConfiguration(configuration);
+		final ResourceManagerRuntimeServicesConfiguration rmServicesConfiguration = ResourceManagerRuntimeServicesConfiguration.fromConfiguration(configuration);
+		final ResourceManagerRuntimeServices rmRuntimeServices = ResourceManagerRuntimeServices.fromConfiguration(
+			rmServicesConfiguration,
+			highAvailabilityServices,
+			rpcService.getScheduledExecutor());
+
+		return new YarnResourceManager(
+			rpcService,
+			ResourceManager.RESOURCE_MANAGER_NAME,
+			resourceId,
+			configuration,
+			System.getenv(),
+			rmConfiguration,
+			highAvailabilityServices,
+			heartbeatServices,
+			rmRuntimeServices.getSlotManager(),
+			metricRegistry,
+			rmRuntimeServices.getJobLeaderIdService(),
+			fatalErrorHandler);
+	}
+
+	@Override
+	protected JobGraph retrieveJobGraph(Configuration configuration) throws FlinkException {
+		String jobGraphFile = configuration.getString(JOB_GRAPH_FILE_PATH, "job.graph");
```

Comment:
Ideally a standard approach to locating the job to launch could be developed across K8/YARN/etc.   It hadn't occurred to me that the docker image would have a serialized job graph within it (as opposed to the packaged program w/ `main` method) , but it definitely simplifies the recovery model.    This has me wondering about how `flink run` would combine with this.